File: src/core/units.ts
/**
 * Canonical units core (dashboard task #19).
 *
 * The data plane keeps amplitudes in ONE canonical unit — RMS volts (Vrms) —
 * and this module converts a canonical value to whatever unit a graph/axis or
 * readout wants to show: peak volts, dBV, dBu, dBFS, dBr, watts, or percent.
 *
 * Keeping a single canonical unit means a trace is measured once and displayed
 * in any unit without re-measuring, and two traces in different units can share
 * a graph by projecting both through here. All functions are pure (no DOM), so
 * they are unit-tested directly.
 *
 * References:
 *   dBV  → 1 Vrms
 *   dBu  → 0.7746 Vrms (√(1 mW into 600 Ω))
 *   dBFS → the device's full-scale Vrms for the active input range (in refs)
 *   dBr  → a user/reference level in Vrms (in refs)
 *   W    → v² / load Ω (in refs)
 *   %    → v / reference Vrms × 100 (in refs)
 */

import type { FdUnit, TdUnit, Unit } from "./model";

/** Reference levels a conversion may need (device/range/context dependent). */
export interface UnitRefs {
  /** Vrms that corresponds to 0 dBFS for the active input range. */
  fullScaleVrms: number;
  /** Reference level for dBr / percent (e.g. a stored "0 dBr" level). */
  refVrms: number;
  /** Load resistance (Ω) used to turn volts into watts. */
  loadOhms: number;
}

/** Neutral defaults (1 Vrms full-scale & reference, 8 Ω load). */
export const DEFAULT_REFS: UnitRefs = { fullScaleVrms: 1, refVrms: 1, loadOhms: 8 };

const DBV_REF = 1; // 1 Vrms
const DBU_REF = 0.7745966692414834; // √(0.001 W · 600 Ω)
const SQRT2 = Math.SQRT2;

/** dB-family (logarithmic) units — displayed to 0.1 dB, floor at −∞. */
const DB_UNITS: ReadonlySet<Unit> = new Set(["dbv", "dbu", "dbfs", "dbr", "db"]);

export function isDbUnit(unit: Unit): boolean {
  return DB_UNITS.has(unit);
}

/** Amplitude ratio → dB (20·log10); −∞ for silence. */
export const db = (ratio: number): number => (ratio > 0 ? 20 * Math.log10(ratio) : -Infinity);
/** dB → amplitude ratio (inverse of {@link db}). */
export const undb = (value: number): number => Math.pow(10, value / 20);

/** The constant dB offset between the dBu and dBV scales (≈ +2.2185 dB):
 * `dBu = dBV + DBU_OVER_DBV_DB`. Single TS definition — the Rust twin is
 * `measurements::units::dbv_to_dbu_db()`. */
export const DBU_OVER_DBV_DB = 20 * Math.log10(DBV_REF / DBU_REF);

/* ------------------------------------------------------------------ */
/* Per-converter display mapping (the #51/#60 pair, in one place).      */
/* ------------------------------------------------------------------ */

/** dB to add to a trace's raw (own-converter dBFS) spectrum for `unit`. */
export function displayOffsetDb(unit: FdUnit, traceOffsetDb: number | null): number {
  switch (unit) {
    case "dbfs":
      return 0;
    case "dbv":
      return traceOffsetDb ?? 0;
    case "dbu":
      return (traceOffsetDb ?? 0) + DBU_OVER_DBV_DB;
  }
}

/** Linear factor mapping a full-scale sample of the trace's converter to
 * `unit` — the td twin of {@link displayOffsetDb} (#60). */
export function displayScale(unit: TdUnit, traceOffsetDb: number | null): number {
  switch (unit) {
    case "pctfs":
      return 100;
    case "v":
      return Math.pow(10, (traceOffsetDb ?? 0) / 20);
    case "mv":
      return 1000 * Math.pow(10, (traceOffsetDb ?? 0) / 20);
  }
}

/** Convert a canonical Vrms amplitude to a target unit's numeric value. */
export function fromVrms(vrms: number, unit: Unit, refs: UnitRefs = DEFAULT_REFS): number {
  const v = Math.max(0, vrms);
  switch (unit) {
    case "vrms":
      return v;
    case "vpk":
      return v * SQRT2;
    case "dbv":
    case "db":
      return db(v / DBV_REF);
    case "dbu":
      return db(v / DBU_REF);
    case "dbfs":
      return db(v / refs.fullScaleVrms);
    case "dbr":
      return db(v / refs.refVrms);
    case "percent":
      return refs.refVrms > 0 ? (v / refs.refVrms) * 100 : 0;
    case "watt":
      return refs.loadOhms > 0 ? (v * v) / refs.loadOhms : 0;
  }
}

/** Inverse of {@link fromVrms}: a unit's numeric value back to canonical Vrms. */
export function toVrms(value: number, unit: Unit, refs: UnitRefs = DEFAULT_REFS): number {
  switch (unit) {
    case "vrms":
      return value;
    case "vpk":
      return value / SQRT2;
    case "dbv":
    case "db":
      return DBV_REF * undb(value);
    case "dbu":
      return DBU_REF * undb(value);
    case "dbfs":
      return refs.fullScaleVrms * undb(value);
    case "dbr":
      return refs.refVrms * undb(value);
    case "percent":
      return (refs.refVrms * value) / 100;
    case "watt":
      return Math.sqrt(Math.max(0, value) * refs.loadOhms);
  }
}

/** Short axis/badge caption for a unit (e.g. "dBV", "Vrms", "%"). */
export function unitLabel(unit: Unit): string {
  switch (unit) {
    case "vrms":
      return "Vrms";
    case "vpk":
      return "Vpk";
    case "dbv":
      return "dBV";
    case "dbu":
      return "dBu";
    case "dbfs":
      return "dBFS";
    case "dbr":
      return "dBr";
    case "watt":
      return "W";
    case "percent":
      return "%";
    case "db":
      return "dB";
  }
}

/** Adaptive SI prefix for a linear magnitude (V or W): value scaled + prefix. */
function siScale(x: number): { value: number; prefix: string } {
  const a = Math.abs(x);
  if (a === 0 || !isFinite(a)) return { value: x, prefix: "" };
  if (a >= 1) return { value: x, prefix: "" };
  if (a >= 1e-3) return { value: x * 1e3, prefix: "m" };
  if (a >= 1e-6) return { value: x * 1e6, prefix: "µ" };
  return { value: x * 1e9, prefix: "n" };
}

export interface FormatOpts {
  /** Significant decimals for dB / percent readouts (default 1). */
  dbDigits?: number;
  /** Significant figures for linear (V / W) readouts (default 3). */
  sig?: number;
}

/**
 * Format a canonical Vrms amplitude in a unit as a display string with its
 * suffix (e.g. "−6.0 dBV", "1.23 Vrms", "12.3 mVrms", "0.50 W", "50.0 %").
 * dB-family units below −200 dB show "−∞".
 */
export function formatVrms(
  vrms: number,
  unit: Unit,
  refs: UnitRefs = DEFAULT_REFS,
  opts: FormatOpts = {}
): string {
  const value = fromVrms(vrms, unit, refs);
  const label = unitLabel(unit);
  if (isDbUnit(unit)) {
    if (!isFinite(value) || value <= -200) return `−∞ ${label}`;
    return `${minus(value.toFixed(opts.dbDigits ?? 1))} ${label}`;
  }
  if (unit === "percent") {
    return `${minus(value.toFixed(opts.dbDigits ?? 1))} ${label}`;
  }
  // Linear volts / watts: adaptive SI prefix + ~3 significant figures.
  const { value: scaled, prefix } = siScale(value);
  const sig = opts.sig ?? 3;
  const intDigits = scaled === 0 ? 1 : Math.max(1, Math.floor(Math.log10(Math.abs(scaled))) + 1);
  const digits = Math.max(0, sig - intDigits);
  // "Vrms"/"Vpk" keep their rms/pk hint after the SI prefix (e.g. "mVrms").
  const suffix =
    unit === "watt" ? `${prefix}W` : `${prefix}V${label.startsWith("V") ? label.slice(1) : ""}`;
  return `${minus(scaled.toFixed(digits))} ${suffix}`;
}

/** Normalise a leading ASCII hyphen to the app's U+2212 minus sign. */
function minus(s: string): string {
  return s.startsWith("-") ? `−${s.slice(1)}` : s;
}

File: src-tauri/src/measurements/units.rs
//! Canonical unit conversions — the Rust twin of `src/dashboard/units.ts`.
//!
//! The data plane keeps amplitudes in ONE canonical unit — RMS volts (Vrms) —
//! and this module converts a canonical value to whatever a readout wants to
//! show: peak volts, dBV, dBu, dBFS, dBr, watts, or percent. Formatting
//! (rounding, SI prefixes, "−∞" rendering) stays in the frontend; only the
//! numeric conversion lives here.
//!
//! References:
//!   dBV  → 1 Vrms
//!   dBu  → 0.7746 Vrms (√(1 mW into 600 Ω))
//!   dBFS → the device's full-scale Vrms for the active range (in [`UnitRefs`])
//!   dBr  → a user/reference level in Vrms (in [`UnitRefs`])
//!   W    → v² / load Ω (in [`UnitRefs`])
//!   %    → v / reference Vrms × 100 (in [`UnitRefs`])

use serde::{Deserialize, Serialize};

/// Floor for every ratio→dB conversion in this crate's measurement layer.
///
/// −200 dB corresponds to an amplitude ratio of 1e-10 — far below both the
/// 24-bit converter noise floor and anything the UI distinguishes from
/// silence (the display layer renders ≤ −200 dB as "−∞"). Flooring here
/// avoids `log(0) = -∞`, which JSON/IPC cannot represent. The same floor was
/// already the de-facto convention in `rest::db` and `audio/weighting.rs`.
pub const DB_FLOOR: f64 = -200.0;

/// 0 dBV reference: 1 Vrms.
pub const DBV_REF_VRMS: f64 = 1.0;

/// 0 dBu reference: √(0.001 W · 600 Ω) ≈ 0.7746 Vrms.
pub const DBU_REF_VRMS: f64 = 0.774_596_669_241_483_4;

/// Amplitude ratio → dB (20·log10), floored at [`DB_FLOOR`].
pub fn db20(ratio: f64) -> f64 {
    if ratio > 0.0 {
        (20.0 * ratio.log10()).max(DB_FLOOR)
    } else {
        DB_FLOOR
    }
}

/// dB → amplitude ratio (inverse of [`db20`]).
pub fn undb20(db: f64) -> f64 {
    10f64.powf(db / 20.0)
}

/// Power ratio → dB (10·log10), floored at [`DB_FLOOR`].
///
/// Distinct from [`db20`]: use this for power-domain quantities (spectrum
/// power averaging, noise power sums), where the ÷10 form applies.
pub fn db10(ratio: f64) -> f64 {
    if ratio > 0.0 {
        (10.0 * ratio.log10()).max(DB_FLOOR)
    } else {
        DB_FLOOR
    }
}

/// dB → power ratio (inverse of [`db10`]).
pub fn undb10(db: f64) -> f64 {
    10f64.powf(db / 10.0)
}

/// The constant dB offset between the dBu and dBV scales:
/// `dBu = dBV + dbv_to_dbu_db()` ≈ +2.2185 dB.
///
/// Single source of truth for a constant that was independently defined three
/// times in the frontend (`levels.ts`, `dashboard/grid.ts`, `units.ts`) and
/// hard-coded once more in `main.ts`.
pub fn dbv_to_dbu_db() -> f64 {
    20.0 * (DBV_REF_VRMS / DBU_REF_VRMS).log10()
}

/// The +3.01 dB between a sine's peak and its RMS: `20·log10(√2)`.
///
/// Used by the DAC-side dBFS→dBV projection: an output range's dBV rating is
/// the RMS of a full-scale *sine*, so the DAC offset carries this term on top
/// of the range (measured convention, task #48).
pub fn sine_peak_to_rms_db() -> f64 {
    20.0 * std::f64::consts::SQRT_2.log10()
}

/// A display/interchange unit for a canonical Vrms amplitude.
///
/// Serialized names match the frontend's `Unit` string union
/// (`"vrms" | "vpk" | "dbv" | ...`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Unit {
    Vrms,
    Vpk,
    Dbv,
    Dbu,
    Dbfs,
    Dbr,
    Percent,
    Watt,
    /// Bare relative dB (same scale as dBV; used for gain/transfer axes).
    Db,
}

impl Unit {
    /// Whether the unit is logarithmic (dB-family).
    pub fn is_db(self) -> bool {
        matches!(self, Unit::Dbv | Unit::Dbu | Unit::Dbfs | Unit::Dbr | Unit::Db)
    }
}

/// Reference levels a conversion may need (device/range/context dependent).
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnitRefs {
    /// Vrms that corresponds to 0 dBFS for the active converter range.
    pub full_scale_vrms: f64,
    /// Reference level for dBr / percent (e.g. a stored "0 dBr" level).
    pub ref_vrms: f64,
    /// Load resistance (Ω) used to turn volts into watts.
    pub load_ohms: f64,
}

impl Default for UnitRefs {
    /// Neutral defaults (1 Vrms full-scale & reference, 8 Ω load) — mirrors
    /// the frontend's `DEFAULT_REFS`.
    fn default() -> Self {
        UnitRefs { full_scale_vrms: 1.0, ref_vrms: 1.0, load_ohms: 8.0 }
    }
}

/// Convert a canonical Vrms amplitude to a target unit's numeric value.
///
/// Negative inputs clamp to 0 (an RMS amplitude cannot be negative); dB-family
/// outputs floor at [`DB_FLOOR`].
pub fn from_vrms(vrms: f64, unit: Unit, refs: &UnitRefs) -> f64 {
    let v = vrms.max(0.0);
    match unit {
        Unit::Vrms => v,
        Unit::Vpk => v * std::f64::consts::SQRT_2,
        Unit::Dbv | Unit::Db => db20(v / DBV_REF_VRMS),
        Unit::Dbu => db20(v / DBU_REF_VRMS),
        Unit::Dbfs => db20(v / refs.full_scale_vrms),
        Unit::Dbr => db20(v / refs.ref_vrms),
        Unit::Percent => {
            if refs.ref_vrms > 0.0 {
                v / refs.ref_vrms * 100.0
            } else {
                0.0
            }
        }
        Unit::Watt => {
            if refs.load_ohms > 0.0 {
                v * v / refs.load_ohms
            } else {
                0.0
            }
        }
    }
}

/// Inverse of [`from_vrms`]: a unit's numeric value back to canonical Vrms.
pub fn to_vrms(value: f64, unit: Unit, refs: &UnitRefs) -> f64 {
    match unit {
        Unit::Vrms => value,
        Unit::Vpk => value / std::f64::consts::SQRT_2,
        Unit::Dbv | Unit::Db => DBV_REF_VRMS * undb20(value),
        Unit::Dbu => DBU_REF_VRMS * undb20(value),
        Unit::Dbfs => refs.full_scale_vrms * undb20(value),
        Unit::Dbr => refs.ref_vrms * undb20(value),
        Unit::Percent => refs.ref_vrms * value / 100.0,
        Unit::Watt => (value.max(0.0) * refs.load_ohms).sqrt(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const ALL_UNITS: [Unit; 9] = [
        Unit::Vrms,
        Unit::Vpk,
        Unit::Dbv,
        Unit::Dbu,
        Unit::Dbfs,
        Unit::Dbr,
        Unit::Percent,
        Unit::Watt,
        Unit::Db,
    ];

    fn refs() -> UnitRefs {
        UnitRefs::default()
    }

    #[test]
    fn vrms_passes_through_and_vpk_scales_by_sqrt2() {
        assert_eq!(from_vrms(1.0, Unit::Vrms, &refs()), 1.0);
        assert!((from_vrms(1.0, Unit::Vpk, &refs()) - std::f64::consts::SQRT_2).abs() < 1e-12);
    }

    #[test]
    fn dbv_reference_points() {
        assert!(from_vrms(1.0, Unit::Dbv, &refs()).abs() < 1e-9);
        assert!((from_vrms(2.0, Unit::Dbv, &refs()) - 6.0206).abs() < 1e-3);
        assert!((from_vrms(0.5, Unit::Dbv, &refs()) + 6.0206).abs() < 1e-3);
    }

    #[test]
    fn dbu_references_0_7746_vrms() {
        assert!(from_vrms(DBU_REF_VRMS, Unit::Dbu, &refs()).abs() < 1e-9);
        assert!((from_vrms(1.0, Unit::Dbu, &refs()) - 2.2185).abs() < 1e-3);
        assert!((dbv_to_dbu_db() - 2.2185).abs() < 1e-3);
    }

    #[test]
    fn dbfs_references_the_full_scale_vrms() {
        let r = UnitRefs { full_scale_vrms: 2.0, ..refs() };
        assert!(from_vrms(2.0, Unit::Dbfs, &r).abs() < 1e-9);
        assert!((from_vrms(1.0, Unit::Dbfs, &r) + 6.0206).abs() < 1e-3);
    }

    #[test]
    fn percent_is_a_linear_ratio_to_the_reference() {
        assert!((from_vrms(0.5, Unit::Percent, &refs()) - 50.0).abs() < 1e-9);
        assert!((from_vrms(1.0, Unit::Percent, &refs()) - 100.0).abs() < 1e-9);
    }

    #[test]
    fn watts_is_v_squared_over_load() {
        assert!((from_vrms(2.0, Unit::Watt, &refs()) - 0.5).abs() < 1e-9);
        assert!((from_vrms(8f64.sqrt(), Unit::Watt, &refs()) - 1.0).abs() < 1e-9);
    }

    #[test]
    fn clamps_negative_volts_and_floors_db_at_silence() {
        assert_eq!(from_vrms(-1.0, Unit::Vrms, &refs()), 0.0);
        // TS returned -Infinity here; the agreed canonical form floors at -200
        // (the display layer shows anything <= -200 as "−∞").
        assert_eq!(from_vrms(0.0, Unit::Dbv, &refs()), DB_FLOOR);
        assert_eq!(db20(0.0), DB_FLOOR);
        assert_eq!(db10(-1.0), DB_FLOOR);
    }

    #[test]
    fn round_trips_all_units() {
        let r = UnitRefs { full_scale_vrms: 3.1, ref_vrms: 0.5, load_ohms: 4.0 };
        for unit in ALL_UNITS {
            for v in [0.001, 0.05, 0.5, 1.0, 2.5] {
                let rt = to_vrms(from_vrms(v, unit, &r), unit, &r);
                assert!((rt - v).abs() < 1e-9, "{unit:?} {v} -> {rt}");
            }
        }
    }

    #[test]
    fn power_db_uses_the_div10_form() {
        assert!((db10(0.1) + 10.0).abs() < 1e-9);
        assert!((undb10(-10.0) - 0.1).abs() < 1e-12);
        assert!((undb20(-20.0) - 0.1).abs() < 1e-12);
    }

    #[test]
    fn sine_peak_to_rms_is_3_01_db() {
        assert!((sine_peak_to_rms_db() - 3.0103).abs() < 1e-3);
    }

    #[test]
    fn unit_serde_names_match_the_frontend() {
        assert_eq!(serde_json::to_string(&Unit::Dbfs).unwrap(), "\"dbfs\"");
        assert_eq!(serde_json::to_string(&Unit::Vrms).unwrap(), "\"vrms\"");
        assert_eq!(
            serde_json::from_str::<Unit>("\"percent\"").unwrap(),
            Unit::Percent
        );
        assert!(Unit::Dbr.is_db());
        assert!(!Unit::Watt.is_db());
    }
}
